AppCenterDistribute/AppCenterDistribute/MSDistribute.m

```diff
@@ -561,70 +558,41 @@ - (void)openUrlInAuthenticationSessionOrSafari:(NSURL *)url {
-        return;
-      }
-      strongSelf.authenticationSession = nil;
-      if (error != nil) {
-        MSLogDebug([MSDistribute logTag], @"Called %@ with error: %@", callbackUrl, error.localizedDescription);
-      }
-
-      // This is error.code == SFAuthenticationErrorCanceledLogin which we can't use to retain backward compatibility.
-      if (error.code == 1) {
-        MSLogError([MSDistribute logTag], @"Authentication session was cancelled by user or failed.");
-      }
-      if (callbackUrl) {
-        [strongSelf openURL:callbackUrl];
-      }
-    };
-
-    // Initialize the SFAuthenticationSession.
-    typedef void (*MSInitSFAuthenticationSession)(id, SEL, NSURL *, NSString *, MSCompletionBlockForAuthSession);
-    MSInitSFAuthenticationSession initMethodCall;
-    initMethodCall = (MSInitSFAuthenticationSession)[session methodForSelector:initSelector];
-    initMethodCall(session, initSelector, url, callbackUrlScheme, authCompletionBlock);
-
-    // Retain the session.
-    self.authenticationSession = session;
-
-    // Create selector for [instanceOfSFAuthenticationSession start].
-    SEL startSelector = NSSelectorFromString(@"start");
+  // The completion block that we need to invoke.
+  __weak typeof(self) weakSelf = self;
+  typedef void (^MSCompletionBlockForAuthSession)(NSURL *callbackUrl, NSError *error);
+  MSCompletionBlockForAuthSession authCompletionBlock = ^(NSURL *callbackUrl, NSError *error) {
+    typeof(self) strongSelf = weakSelf;
+    if (!strongSelf) {
+      return;
+    }
+    strongSelf.authenticationSession = nil;
+    if (error) {
+      MSLogDebug([MSDistribute logTag], @"Called %@ with error: %@", callbackUrl, error.localizedDescription);
+    }
+    if (error.code == SFAuthenticationErrorCanceledLogin) {
+      MSLogError([MSDistribute logTag], @"Authentication session was cancelled by user or failed.");
+    }
+    if (callbackUrl) {
+      [strongSelf openURL:callbackUrl];
+    }
+  };
+  SFAuthenticationSession *session = [[SFAuthenticationSession alloc] initWithURL:url callbackURLScheme:callbackUrlScheme completionHandler:authCompletionBlock];
 
-    // Call [SFAuthenticationSession start] dynamically.
-    typedef BOOL (*MSStartSFAuthenticationSession)(id, SEL);
-    MSStartSFAuthenticationSession startMethodCall;
-    startMethodCall = (MSStartSFAuthenticationSession)[session methodForSelector:startSelector];
-    BOOL success = startMethodCall(session, startSelector);
+  // Retain the session.
+  self.authenticationSession = session;
+  @try {
+    BOOL success = [session start];
     if (success) {
-      MSLogDebug([MSDistribute logTag], @"Authentication Session Started, showing confirmation dialog");
+      MSLogDebug([MSDistribute logTag], @"Authentication session started, showing confirmation dialog.");
     }
-#pragma clang diagnostic pop
+  } @catch (NSException *exception) {
+    MSLogError([MSDistribute logTag], @"Failed to start authentication session: %@", exception);
```

Comment:
Should we print the reason? Because we already have the class, method and line directly in the console log.
```suggestion
    MSLogError([MSDistribute logTag], @"Failed to start authentication session: %@", exception.reason);
```